// react-native-fbsdk-next/lib/module/FBGraphRequest.js
function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

/**
 * Copyright (c) 2015-present, Facebook, Inc. All rights reserved.
 *
 * You are hereby granted a non-exclusive, worldwide, royalty-free license to use,
 * copy, modify, and distribute this software in source code or binary form for use
 * in connection with the web services and APIs provided by Facebook.
 *
 * As with any software that integrates with the Facebook platform, your use of
 * this software is subject to the Facebook Developer Principles and Policies
 * [http://developers.facebook.com/policy/]. This copyright notice shall be
 * included in all copies or substantial portions of the software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS
 * FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR
 * COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER
 * IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN
 * CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
 *
 * @format
 */

/**
 * Represents a Graph API request and provides batch request supports.
 */
class FBGraphRequest {
  /**
   * The Graph API endpoint to use for the request, for example "me".
   */

  /**
   * The optional config for the request.
   */

  /**
   * Called upon completion or failure of the request.
   */

  /**
   * Constructs a new Graph API request.
   */
  constructor(graphPath, config, callback) {
    _defineProperty(this, "graphPath", void 0);

    _defineProperty(this, "config", void 0);

    _defineProperty(this, "callback", void 0);

    this.graphPath = graphPath;
    this.config = config ? config : {};
    this.callback = callback ? callback : () => null;
  }
  /**
   * Adds a string parameter to the request.
   */


  addStringParameter(paramString, key) {
    if (this.config != null && this.config.parameters != null) {
      this.config.parameters[key] = {
        string: paramString
      };
    }
  }

}

export default FBGraphRequest;
